InfiniTAM/Engine/RosPoseSourceEngine.cpp

```diff
@@ -0,0 +1,191 @@
+// Copyright 2014-2015 Isis Innovation Limited and the authors of InfiniTAM
+
+#include "RosPoseSourceEngine.h"
+#include <glog/logging.h>
+
+#include <cstdio>
+#include <stdexcept>
+#include <string>
+
+#include "../Utils/FileUtils.h"
+
+#ifdef COMPILE_WITH_Ros
+
+namespace InfiniTAM {
+namespace Engine {
+
+RosPoseSourceEngine::RosPoseSourceEngine(ros::NodeHandle& nh)
+    : PoseSourceEngine(),
+      first_time_tf_available_(true),
+      broadcast_transformations(false) {
+  set_camera_pose_ = true;
+
+  // set the topic and frame_id names.
+  nh.param<std::string>("camera_frame_id", camera_frame_id_,
+                        "sr300_depth_optical_frame");
+  nh.param<std::string>("world_frame_id", world_frame_id_, "world");
+  nh.param<std::string>("complete_cloud", complete_cloud_topic_,
+                        "/complete_cloud");
+}
+
+RosPoseSourceEngine::~RosPoseSourceEngine() {
+}
+
+// Get the pose of the camera from the forward kinematics of the robot.
+void RosPoseSourceEngine::TFCallback(const tf::tfMessage& tf_msg) {
+  try {
+    // get transform from world to camera frame.
+    listener.lookupTransform(world_frame_id_, camera_frame_id_, ros::Time(0),
```

Comment:
Make sure you use the proper ros times.
You are usually interested in the timestamps when the camera frame is stored, maybe set this as a member in the main_engine or somewhere where you have access to it from both ros engines.